# **Module 2 — Data Cleaning, Analysis, Feature Engineernig**

# **Module 2 Part 1**

# 1. How to **Load pandas library and read dataset**

In [1]:
import pandas as pd
df = pd.read_csv("../01_Data/Titanic.csv")
display(df.head())

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# 2. How to **Check all null values**

In [3]:
# find missing values(NAN)
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# **Observation**
*(**In this dataset, the `Age` column has 177 missing values, the `Cabin` column has 687 missing values, and the `Embarked` column has 2 missing values.**)*

# 3. How to **Handle Missing Values**

### *Method  — fillna()*
*Fill Age column NaN values by its median because Age is a numerical column.*


### Missing Values Handling
**We use different methods depending on the type of data:**

*- Numeric column: median/mean*

*- Categorical column: mode*

*- If many values are missing, use a custom label like "Missing" or "Unknown"*


In [4]:

# 1) Age: numeric column
df["Age"] = df["Age"].fillna(df["Age"].median())

# 2) Embarked: categorical column
mode_value = df["Embarked"].mode()
if not mode_value.empty:
    df["Embarked"] = df["Embarked"].fillna(mode_value.iloc[0])
else:
    df["Embarked"] = df["Embarked"].fillna("Unknown")

# 3) Cabin: many missing values, so use custom label
df["Cabin"] = df["Cabin"].fillna("Value_Missing")


In [5]:
# Check null values again for cheking function is properly works or not
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

# 4. How to Perform **dropna()**

*It removes every row that contains even one `NaN` value. It returns a new DataFrame, so if I want to save the changes permanently, I need to assign it back to `df` or use `inplace=True` or `df = df.dropna()`.*

In [6]:
df.dropna()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,Value_Missing,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,Value_Missing,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,Value_Missing,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,Value_Missing,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,Value_Missing,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


# 5. How to Perform **Data Type Conversion**

In [7]:
# Check Datatype first
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [12]:
# numeric conversion
df["Age"] = df["Age"].astype(int)
df["Fare"] = df["Fare"].astype(int)

# Categorical conversion
df["Sex"] = df["Sex"].astype("category")
df["Embarked"] = df["Embarked"].astype("category")

# Integer conversion
df["Survived"] = df["Survived"].astype(float)
df["Pclass"] = df["Pclass"].astype(float)

print(df.dtypes)

PassengerId       int64
Survived        float64
Pclass          float64
Name             object
Sex            category
Age               int64
SibSp             int64
Parch             int64
Ticket           object
Fare              int64
Cabin            object
Embarked       category
dtype: object


### **df["--"] = pd.to_numeric(df["--"], errors="coerce")**

*If a value cannot be converted into a number, it will be changed to `NaN` instead of producing an error. To save this change permanently, use:*

In [13]:
df["Age"] = pd.to_numeric(df["Age"], errors="coerce")
print(df["Age"])

0      22
1      38
2      26
3      35
4      35
       ..
886    27
887    19
888    28
889    26
890    32
Name: Age, Length: 891, dtype: int64


# 6. How to use **Statistics**
### Analyze dataset

Mean

In [14]:
df["Age"].mean()

np.float64(29.34567901234568)

Median

In [15]:
df["Age"].median()

np.float64(28.0)

standard Deviation

In [16]:
df["Age"].std()

np.float64(13.02821186279203)

Mode

In [17]:
df["Sex"].mode()

0    male
Name: Sex, dtype: category
Categories (2, object): ['female', 'male']

Minimum

In [18]:
df["PassengerId"].min()

np.int64(1)

Maximum


In [19]:
df["PassengerId"].max()

np.int64(891)

In [20]:
# All statistics summary
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.345679,0.523008,0.381594,31.785634
std,257.353842,0.486592,0.836071,13.028212,1.102743,0.806057,49.703730
min,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,22.000000,0.000000,0.000000,7.000000
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.000000
75%,668.500000,1.000000,3.000000,35.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.000000


# 7. How to use **unique()**

**Which distinct values are present in a column?**

*Use `unique()` when you want to find the possible categories in a categorical column.*

In [21]:
df["Pclass"].unique()

array([3., 1., 2.])

# 8. How to use **value_count()**
### Find how many times a value appears in a column.

In [22]:
# df["Age"].value_counts()
df["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

### Defference between .unique() & .value_count()

`unique()` = This method displays the different values present in a column.

`value_counts()` = This method counts each distinct value in a column and shows how many times it appears.

### **Observation of Day 3**
*I check for null values using `df.isnull().sum()`.*

*I fill missing values in numerical columns with the median and missing values in categorical columns with the mode.*

*For example:*

`df["Age"] = df["Age"].fillna(df["Age"].median())`

*I use the median because `Age` is a numerical column. For the categorical `Embarked` column, I use:*
`df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])`

*If a column contains many missing values, such as `Cabin`, where 687 out of 891 values are missing, I can use the `dropna()` function to remove rows with missing cabin data. However, I should treat this as a warning because the `Cabin` column may be useful for passenger analysis, so I should not delete it without careful consideration.*

*I can calculate individual statistics for a column. For example, to find the standard deviation of the `Age` column, I use:*

`df["Age"].std()`

*Similarly, I can calculate other statistical values, such as the minimum, maximum, mean, median, and mode. I can also use the `df.describe()` method to find several statistics for all numerical columns at once, including count, mean, standard deviation, minimum, maximum, and the 25%, 50%, and 75% percentiles. Before handling null values, `describe()` also shows the count of non-null values.*
*The `df["Column"].unique()` method displays the different values present in a column. The `df["Column"].value_counts()` method counts how many times each unique value appears in that column.*

# **Module 2 Part 2 — Analysis + Transformation**

# 9A .How to use **Single_groupby()**
### *This is an important Pandas concept: finding the average fare for male and female passengers.*

In [23]:
df.groupby("Sex", observed=False)["Fare"].mean()

Sex
female    44.063694
male      25.103986
Name: Fare, dtype: float64

### Logic Behind `groupby()`

Create groups based on `Sex`  
↓  
Male  
Female  
↓  
Select the `Fare` values from each group  
↓  
Calculate the average fare

# 9B. How to use **Multiple Groupby**


In [24]:
df.groupby(["Sex", "Pclass"], observed=False)["Fare"].mean()

Sex     Pclass
female  1.0       105.691489
        2.0        21.776316
        3.0        15.597222
male    1.0        66.836066
        2.0        19.527778
        3.0        12.167147
Name: Fare, dtype: float64

# 10. How to use **Miltiple Aggregations**
agar Average age , Maximum Age, Average Fare

In [25]:
df.groupby("Sex", observed=False).agg({
    "Age": "mean",
    "Fare": "mean"
})

,Age,Fare
Sex,,
female,27.920382,44.063694
male,30.121317,25.103986


In [26]:
df.groupby("Sex", observed=False).agg({
    "Age": ["mean", "max"],
    "Fare": ["mean", "max"]
})

Age           Fare     
             mean max       mean  max
Sex                                  
female  27.920382  63  44.063694  512
male    30.121317  80  25.103986  512

### My Self-Learning About `groupby()`

***Through self-learning, I discovered that the `groupby()` method treats object and categorical data types differently. When a categorical column is used, the code runs successfully but may display a `FutureWarning`:**,*

***`The default of observed=False is deprecated and will be changed to True in a future version of pandas.`**,*

***To remove this warning, I asked the AI assistant in VS Code for help. I learned how Pandas handles object and categorical data types differently when using `groupby()`.**,*
***At first, I thought about changing the data type of the `Sex` column. However, I learned that adding `observed=False` makes the code explicit:**,*

***`df.groupby("Sex", observed=False)`**,*
***This removes the warning while allowing me to continue using the efficiency of the categorical data type.**,*

### **This was my self-learning about `groupby()`.**

**----------------------------------------------------------------------------------------------------------------------------------------------------------------**

# 11. How to **Create New Columns/Feature Engineering**
### Feature Engineering = Create useful features(columns) from existing columns
In the Titanic dataset:

SibSp = Siblings/Spouse, Parch = Parents/Children

Something like this can be created from both:

     New column   =   existing columns
### [FamilySize = SibSp + Parch + 1]
+1 because the passenger is also a member of the family.


In [27]:
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1
df["FamilySize"]

0      2
1      2
2      1
3      2
4      1
      ..
886    1
887    1
888    4
889    1
890    1
Name: FamilySize, Length: 891, dtype: int64

# 12. Creating Age Group/Feature Engineering
### we are creating category

In [28]:
df["AgeGroup"] = pd.cut(
    df["Age"],
    bins = [0, 18, 60, 100],
    labels = ["Child", "Adult", "Senior"]
)
df["AgeGroup"]

0      Adult
1      Adult
2      Adult
3      Adult
4      Adult
       ...  
886    Adult
887    Adult
888    Adult
889    Adult
890    Adult
Name: AgeGroup, Length: 891, dtype: category
Categories (3, object): ['Child' < 'Adult' < 'Senior']

# 13. How to use **rename()**
Change name of a column


In [29]:
df.rename(columns={"Sex": "Gender"}, inplace=True)
df["Gender"]

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Gender, Length: 891, dtype: category
Categories (2, object): ['female', 'male']

# 14. How to use **Cleaning**
### After performing all the operations, it is necessary to give the dataset a final look:

#### 1. Check Duplicate Rows 

In [30]:
df.duplicated().sum()

np.int64(0)

### 2. Remove Duplicates

In [31]:
df = df.drop_duplicates()

### 3. Check Column names 

In [32]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Gender', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'FamilySize',
       'AgeGroup'],
      dtype='object')

### 4. Extra Space Cleaning

In [33]:
df["Name"] = df["Name"].str.strip()

### 5. If all column names need to be converted to lowercase:

In [34]:
df.columns = df.columns.str.lower()

In [36]:
df.to_csv("../04_Output/New_titanic.csv", index=False)
